NetworkResourceLoader::isMainFrameLoad() should be correct with site isolation enabled

achristensen07 opened the pull request:
#### 3479e44a7b3628dd55a1df3a7f5b7919be913f6e
<pre>
NetworkResourceLoader::isMainFrameLoad() should be correct with site isolation enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=266658">https://bugs.webkit.org/show_bug.cgi?id=266658</a>

Reviewed by Pascoe.

frameAncestorOrigins was empty so isMainFrameLoad() was incorrectly returning true for
non-main frames, causing them not to commit loads if cross-origin-opener-policy header
fields were present in the response of an iframe.  Normally,
NetworkResourceLoader::doCrossOriginOpenerHandlingOfResponse would return early because
it wasn&apos;t a main frame load.  That happens again after this PR.
Since a web content process shouldn&apos;t know the origins of ancestor frames that aren&apos;t
same-site, I add nullptr to the Vector&lt;RefPtr&lt;SecurityOrigin&gt;&gt; if I encounter a RemoteFrame
in the tree ancestry.  I added a few previously-unneeded null checks to prevent it from
crashing.

* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::checkFrameAncestors):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::shouldInterruptLoadForXFrameOptions):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.mm:
(WebKit::SubFrameSOAuthorizationSession::shouldInterruptLoadForXFrameOptions):
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::scheduleLoadFromNetworkProcess):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/272306@main">https://commits.webkit.org/272306@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e6e52b7fb4bb05582743512fa834e5577d599bdd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31291 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9961 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32983 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33791 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/28380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32058 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12304 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7213 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28025 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31627 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8388 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/27947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7208 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/7376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/27846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35134 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28454 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/28301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/33510 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7438 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/5481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/31349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/9113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/8137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4073 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/7961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -627,10 +627,10 @@ bool NetworkResourceLoader::shouldInterruptLoadForXFrameOptions(const String& xF
         auto origin = SecurityOrigin::create(url);
         auto topFrameOrigin = m_parameters.frameAncestorOrigins.last();
-        if (!origin->isSameSchemeHostPort(*topFrameOrigin))
+        if (topFrameOrigin && !origin->isSameSchemeHostPort(*topFrameOrigin))
```
Is this still a valid check? e.g. will this still prevent a cross-origin website to embed an i-frame with `XFrameOptionsDisposition::SameOrigin`?

If topFrameOrigin is cross-origin, then it'll be nullptr here I think, but we won't ever return true in this function since the cross-origin ancestors are nullptr and we require both not nullptr and `!origin->isSameSchemeHostPort(*topFrameOrigin)` to return true here.

Maybe `!topFrameOrigin || !origin->isSameSchemeHostPort(*topFrameOrigin)` would work?

pascoej commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -627,10 +627,10 @@ bool NetworkResourceLoader::shouldInterruptLoadForXFrameOptions(const String& xF
             return true;
         for (auto& ancestorOrigin : m_parameters.frameAncestorOrigins) {
-            if (!origin->isSameSchemeHostPort(*ancestorOrigin))
+            if (ancestorOrigin && !origin->isSameSchemeHostPort(*ancestorOrigin))
```
ditto

pascoej commented on Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.mm:
```diff
@@ -147,7 +147,7 @@
         auto origin = SecurityOrigin::create(url);
         for (auto& ancestorOrigin : frameAncestorOrigins) {
-            if (!origin->isSameSchemeHostPort(*ancestorOrigin))
+            if (ancestorOrigin && !origin->isSameSchemeHostPort(*ancestorOrigin))
```
ditto

achristensen07 committed:
NetworkResourceLoader::isMainFrameLoad() should be correct with site isolation enabled
https://bugs.webkit.org/show_bug.cgi?id=266658

Reviewed by Pascoe.

frameAncestorOrigins was empty so isMainFrameLoad() was incorrectly returning true for
non-main frames, causing them not to commit loads if cross-origin-opener-policy header
fields were present in the response of an iframe.  Normally,
NetworkResourceLoader::doCrossOriginOpenerHandlingOfResponse would return early because
it wasn't a main frame load.  That happens again after this PR.
Since a web content process shouldn't know the origins of ancestor frames that aren't
same-site, I add nullptr to the Vector<RefPtr<SecurityOrigin>> if I encounter a RemoteFrame
in the tree ancestry.  I added a few previously-unneeded null checks to prevent it from
crashing.

* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::checkFrameAncestors):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::shouldInterruptLoadForXFrameOptions):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.mm:
(WebKit::SubFrameSOAuthorizationSession::shouldInterruptLoadForXFrameOptions):
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::scheduleLoadFromNetworkProcess):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/272306@main
